File: StockBuyAndSell.c
#include <stdio.h>
int maximumProfit(const int* prices, int n) {
    int res = 0; // Biến lưu trữ lợi nhuận tối đa
    // Duyệt qua mảng giá, cộng chênh lệch giữa các giá tăng dần
    for (int i = 1; i < n; i++) {
        if (prices[i] > prices[i - 1]) 
            res += prices[i] - prices[i - 1]; // Cộng lợi nhuận khi giá tăng
    }
    return res; // Trả về lợi nhuận tối đa
}
int main() {
    int n;
    printf("Nhập số lượng giá cổ phiếu: "); scanf("%d", &n);

    if (n < 2) {
        printf("Cần ít nhất 2 giá để tính lợi nhuận.\n");
        return 1; // Thoát chương trình
    }
    int prices[n];
    printf("Nhập giá cổ phiếu (mỗi giá cách nhau bởi khoảng trắng):");

    for (int i = 0; i < n; i++) {
        scanf("%d", &prices[i]);
    }
    int maxProfit = maximumProfit(prices, n);
    printf("Lợi nhuận tối đa có thể đạt được là: %d\n", maxProfit);
    return 0;
}

File: StockBuyAndSell.py
def maximumProfit(prices):
    n = len(prices)  # Độ dài của danh sách giá
    res = 0          # Biến lưu trữ lợi nhuận tối đa
    i = 0            # Chỉ số duyệt danh sách

    while i < n - 1:
        # Tìm local minima (giá thấp nhất trước khi giá tăng)
        while i < n - 1 and prices[i] >= prices[i + 1]:
            i += 1
        lMin = prices[i]
        # Tìm local maxima (giá cao nhất trước khi giá giảm)
        while i < n - 1 and prices[i] <= prices[i + 1]:
            i += 1
        lMax = prices[i]
        # Tính lợi nhuận từ local minima và maxima
        res += (lMax - lMin)
    return res

if __name__ == "__main__":
    try:
        prices = list(map(int, input("Nhập danh sách giá cổ phiếu (các số cách nhau bởi dấu cách): ").split()))
        if len(prices) < 2:
            print("Vui lòng nhập ít nhất 2 giá để tính lợi nhuận.")
        else:
            max_profit = maximumProfit(prices)
            print(f"Lợi nhuận tối đa có thể đạt được là: {max_profit}")
    except ValueError:
        print("Vui lòng chỉ nhập các số nguyên hợp lệ, cách nhau bởi dấu cách.")